### PR TITLE
DefaultWriteBehindProcessor improvement for partial failures at storeAll/deleteAll

### LINKED ISSUE #9733: DefaultWriteBehindProcessor: do not add whole batchMap in failureList in case of failure
when we use storeAll/deleteAll methods in MapStore interface, it is possible that underlying implementation has stored/deleted some entries from the supplied batch and then gets an exception on some particular entry in the middle of the batch. In order to handle this scenario properly the MapStore could remove from the supplied batch entries which were properly stored in the underlying persistent store. The only thing to improve in the DefaultWiteBehindProcessor is to properly synchronise the map passed to processBatch method with the initial batchMap (https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java#L243):

```
            @Override
            public boolean run() throws Exception {
                callBeforeStoreListeners(batchMap.values());
                final Map map = convertToObject(batchMap);
                try {
                    final boolean result = operationType.processBatch(map, mapStore);
                } catch (Exception ex) {
                    // leave in batchMap only entries which were left in the map...
                   throw ex;
                }
                callAfterStoreListeners(batchMap.values());
                return result;
            }
```
I can provide PR for this, if you wish.

Thanks, Denis.

### PR DESCRIPTION
**Pull request by dsukhoroslov**
Fixes #9733

### CODE REVIEW

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
```	Test Result (1 failure / +1)
com.hazelcast.client.heartbeat.ClientHeartbeatTest.testClientEndpointsDelaySeconds_whenHeartbeatResumed```

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by dsukhoroslov** [comment]
@jerrinot, is the following javadoc ok?

```
    /**
     * Stores multiple entries. Implementation of this method can optimize the
     * store operation by storing all entries in one database connection.
     * storeAll is used when writeDelaySeconds is positive (write-behind).
     * If an exception is thrown, the entries will try to be stored one by one using the store() method.
     * 
     * Note: on the retry phase only entries left in the map will be storied one-by-one. In this way a 
     * MapStore implementation can handle partial storeAll cases when some entries were stored successfully
     * before failure happens. Entries removed from the map will be not passed to subsequent call to 
     * store() method any more.
     *
     * @param map map of entries to store
     */
    void storeAll(Map<K, V> map);

    /**
     * Deletes multiple entries from the store.
     * If an exception is thrown the entries will try to be deleted one by one using the delete() method.
     *
     * Note: on the retry phase only entries left in the map will be deleted one-by-one. In this way a 
     * MapStore implementation can handle partial deleteAll cases when some entries were deleted successfully
     * before failure happens. Entries removed from the map will be not passed to subsequent call to 
     * delete() method any more.

     * @param keys the keys of the entries to delete.
     */
    void deleteAll(Collection<K> keys);
```

**Comment by jerrinot** [comment]
@dsukhoroslov: looks good to me, many thanks!

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java [hunk 1]
```diff
@@ -243,14 +244,22 @@ private Map convertToObject(Map<Object, DelayedEntry> batchMap) {
...
                 callBeforeStoreListeners(batchMap.values());
                 final Map map = convertToObject(batchMap);
-                final boolean result = operationType.processBatch(map, mapStore);
+                boolean result;
+                try {
+                    result = operationType.processBatch(map, mapStore);
+                } catch (Exception ex) {
+                    Iterator<Object> keys = batchMap.keySet().iterator();
+                    while (keys.hasNext()) {
+                        if (!map.containsKey(DefaultWriteBehindProcessor.this.toObject(keys.next()))) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java]: `DefaultWriteBehindProcessor.this.` is redundant here

File: hazelcast/src/main/java/com/hazelcast/core/MapStore.java [hunk 2]
```diff
@@ -50,6 +50,11 @@
      * store operation by storing all entries in one database connection.
      * storeAll is used when writeDelaySeconds is positive (write-behind).
      * If an exception is thrown, the entries will try to be stored one by one using the store() method.
+     * 
+     * Note: on the retry phase only entries left in the map will be storied one-by-one. In this way a
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/core/MapStore.java]: typo: "stored" instead of "storied" (same below)

Comment (dsukhoroslov) [inline hazelcast/src/main/java/com/hazelcast/core/MapStore.java]: fixed

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java (hunk 1 above)
Comment (dsukhoroslov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java]: In my IDE (Eclipse 4.5.1) remove of 'this' reference causes error: Cannot make a static reference to the non-static method toObject(Object) from the type AbstractWriteBehindProcessor<DelayedEntry>

Comment (dsukhoroslov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java]: fixed

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
looks good now, can you please squash your commits and update the PR?

**Comment by jerrinot** [comment]
@dsukhoroslov: I can still see 6 distinct commits. 
GitHub offers to squash commit on merging, but I guess it would remove you as an author and I don't want to do that.

**Comment by dsukhoroslov** [comment]
@jerrinot could you merge it as it is? I don't care about the author name, but this squash op.. don't want waste time with it any more, sorry.

**Comment by jerrinot** [comment]
I did not squash it, but rebased on the master -> your are still the author of the changeset, but the ugly merge commit is gone. see https://github.com/hazelcast/hazelcast/commits/master

thanks again for this contribution!

**Comment by dsukhoroslov** [comment]
thanks :)

**Pull request merged**